datafusion/expr/src/type_coercion/functions.rs

```diff
@@ -430,7 +430,7 @@ fn coerced_from<'a>(
         {
             Some(type_into.clone())
         }
-
+        Dictionary(_, _) if matches!(type_from, Utf8) => Some(type_into.clone()),
```

Comment:
I think this currently says it is possible to coerce from Utf8 to any dictionary type (e.g. even `Dictionary(Int64, Int32)`)

I think the `dictionary_coercion` function does something similar here: https://github.com/apache/arrow-datafusion/blob/37b73751b19b82516443579c714b3b5986dac927/datafusion/expr/src/type_coercion/binary.rs#L635-L634

So maybe we need to verify we can coerce the value types, like this?

```suggestion
        Dictionary(_, value_type) if coerce_from(value_type, from_type) => Some(type_into.clone()),
```